### PR TITLE
[Performance][CUDA] Faster CSRToCOO

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
CUB recently implemented a DeviceCopy::Batched algorithm, which can be used to implement the DeviceRunLength::Decode algorithm, which in turn can be used to implement the CSRToCOO for the GPU in a more efficient way for the 64-bit code path. The current implementation suffers from O(E log N) complexity, which this PR reduces to O(E).

My motivation was that CSRToCOO is called inside the labor sampler algorithm, so improving the performance will also improve Labor Sampler and make its complexity truly linear in the number of edges.

My experimental evaluations show that the Labor Sampler runtime consistently improves with this change.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by mfbalin** [comment]
@BarclayII @yaox12, could you take a look? This PR requires using the latest thrust and cub versions.

**Comment by yaox12** [comment]
1. Has this feature been included in any CUB release version?
2. CUB 2.x introduced the libcu++ dependency. Do you know if it's compatible with CUDA 10.2, since DGL still supports it?

**Comment by yaox12** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
@yaox12
1. Not yet, I just recently contributed DeviceCopy to CUB in this [PR](https://github.com/NVIDIA/cub/pull/675), I guess it will be included in the next release. I don't know the timeline of the next release though.
2. It looks like thrust includes cub and libcu++ as submodule dependencies however I don't know if the libcu++ version that comes with thrust is compatible with CUDA 10.2. What I was able to find is that libcu++ was first released for CUDA 10.2. I don't know if they dropped CUDA 10.2 support in later releases.

**Comment by yaox12** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
It looks like the libcudacxx headers can not be found even though I added it to the cuda include directories, what else needs to be done so that those headers can be found?

Error msg: `/root/jenkins/workspace/dgl_PR-5648@2/third_party/thrust/thrust/detail/type_traits.h:27:10: fatal error: cuda/std/type_traits: No such file or directory`

**Comment by yaox12** [comment]
@dgl-bot

**Comment by mfbalin** [comment]
It looks like the following header can not be found this time: `fatal error: cuda_bf16.h: No such file or directory`

@yaox12 Isn't this header supposed to come with CUDA? Do you have any insights?

**Comment by jermainewang** [comment]
@yaox12  Would you like to review the code?

**Comment by czkkkkkk** [comment]
From the above discussion, this PR may depends on the CUDA version. @mfbalin can you keep the current implementation and use the new code when CUDA version is greater than 11.0?

> Not yet, I just recently contributed DeviceCopy to CUB in this https://github.com/NVIDIA/cub/pull/675, I guess it will be included in the next release. I don't know the timeline of the next release though.

Does it mean this PR depends on a unreleased version of CUB? If so, we cannot merge this PR for now.

**Comment by mfbalin** [comment]
The new CUB and Thrust releases require CUDA 11.0 from my understanding. That is why I was waiting for a reply in #4501 to see if CUDA 10.2 support would be maintained or not. Otherwise, additional effort is required to choose the thrust version according to the CUDA version and maintain multiple versions of the code. I created this PR to show that future thrust/CUB versions have benefits for the DGL code base so that maintainers can decide how to proceed. Even if this PR is not merged now, it could be merged in the future.

**Comment by jermainewang** [comment]
@frozenbugs @Rhett-Ying for the CUDA 10.2 issue in CI.

**Comment by czkkkkkk** [comment]
Hi @mfbalin, we do not deprecate CUDA 10.2 for now. Let's keep this PR open and merge it after the deprecation.

**Comment by mfbalin** [comment]
@czkkkkkk CUDA 10.2 is not in the CI anymore if I am not mistaken with #5977. Are we going to wait for months before the next release of thrust/cub or is it possible to merge this PR before then?

**Comment by yaox12** [comment]
@dgl-bot

File: src/array/cuda/csr2coo.cu
Comment (czkkkkkk) [inline src/array/cuda/csr2coo.cu:93]: `max_copy_at_once` is a pretty large number. Why it is necessary?

File: CMakeLists.txt
Comment (czkkkkkk) [inline CMakeLists.txt:53]: Here introduces more dependencies on thrust. Could you try whether you can use the thrust library in CUDA?

Comment (mfbalin) [reply CMakeLists.txt:53]: They come automatically when the thrust submodule in DGL is updated. New versions of thrust come with its own cub and libcudacxx.

File: src/array/cuda/csr2coo.cu
Comment (mfbalin) [reply src/array/cuda/csr2coo.cu:93]: Because the cub algorithm only works with int32_t types when it comes to sizes. If we want to copy more items, then we call the algorithm multiple times.

Comment (mfbalin) [reply src/array/cuda/csr2coo.cu:93]: For example, if we had a graph 3 billion nodes, then the loop would run multiple times.

File: CMakeLists.txt
Comment (mfbalin) [reply CMakeLists.txt:53]: The CUB function I am using in this PR became part of CUB very recently. Thus, it will be available only with cuda 12. Since dgl supports cuda 11, DGL needs to continue to depend on thrust as a submodule.

**Comment by mfbalin** [comment]
I am not sure if this new code path or the existing int32_t specialization is faster. One would need to do a benchmark to see if the cusparse csr2coo  is better or cub is better.
Cusparse code path: https://github.com/dmlc/dgl/blob/078be4369411e72365248409e39a544ed4451e86/src/array/cuda/csr2coo.cu#L44

File: CMakeLists.txt
Comment (czkkkkkk) [reply CMakeLists.txt:53]: I see.

**Review by czkkkkkk — APPROVED** [review]

File: CMakeLists.txt
Comment (frozenbugs) [reply CMakeLists.txt:53]: Can you put a TODO here to remind us remove this once we deprecating CUDA 11?

**Comment by frozenbugs** [comment]
@mfbalin can you help us take a look at the failing test?

**Comment by mfbalin** [comment]
Partition had a minor bug due to the 1 off specification of histogramEven call parameters. Tests should all pass now.

**Comment by mfbalin** [comment]
Actually, we might want to fix this bug in a separate PR so that if there is any problem with the new thrust version (on the off change), we can take this PR back but the bug fix is still stays.

**Comment by frozenbugs** [comment]
> Actually, we might want to fix this bug in a separate PR so that if there is any problem with the new thrust version (on the off change), we can take this PR back but the bug fix is still stays.

Solid judgement :)

**Comment by frozenbugs** [comment]
> Partition had a minor bug due to the 1 off specification of histogramEven call parameters. Tests should all pass now.

Thanks for fixing, any reason this was not exposed without your optimization?

**Comment by mfbalin** [comment]
The histogram has bins, let's say we are computing a histogram for bins with values 0 to 6 (7 bins in total). Without the fix, the bin boundaries were [0, 8) (8 exclusive) vs [0, 7). Documentation says that CUB evenly distributed the values in the range to bins. But 8 doesn't divide 7. So without the fix, with the new update, 0 and 1 was in the first bin. Without the update 6 and 7 were in the last bin (Which didn't expose the bug). I am speculating that this caused a rounding error and a change in the implementation in CUB might have exposed this bug.

**Pull request merged**
